server/athenian/api/internal/miners/github/deployment.py

```diff
@@ -323,6 +334,55 @@ async def mine_deployments(
     return joined
 
 
+async def _invalidate_pdb_facts_for_newer_deploys(
+    notifications: pd.DataFrame,
+    missed_mask: npt.NDArray[bool],
+    account: int,
+    pdb: DatabaseLike,
+) -> set[str]:
+    """Invalidate precomputed facts about deployments newer that deploys not yet precomputed.
+
+    For each of such deployments rows in pdb about the deploy are cleared.
+
+    `missed_mask` is a bitmask relative to `notifications` stating which deployments are
+    not yet precomputed.
+
+    A set with the names of affected precomputed deployments is returned.
+
+    """
+    missed = notifications[missed_mask]
+    not_missed = notifications[~missed_mask]
+
+    to_invalidate = set()
+
+    for env, missed_env_notifications in missed.groupby("environment"):
+        oldest_missed_time = np.min(missed_env_notifications.finished_at)
+
+        to_invalidate_for_env = not_missed[notifications.environment == env][
+            notifications.started_at > oldest_missed_time
+        ].index.values.astype("U")
+
+        to_invalidate.update(to_invalidate_for_env)
```

Comment:
All right, it's time to show some numpy-fu. Grouping with pandas is extremely, obscenely slow in production. It's a nice shortcut to prototype a solution, but the loop should be either vectorized completely or rewritten afterward.

In our case:

```python
finished_ats = notifications[DeploymentNotification.finished_at.name].values
chrono_order = np.argsort(finished_ats)
envs_col = notifications[DeploymentNotification.environment.name].values[chrono_order]
env_order = np.argsort(envs_col, kind="stable")  # stable sort so that we don't lose the chronological order
final_order = chrono_order[env_order]
missed = missed[final_order]
detect = np.diff(missed, prepend=[0])  # -1 for out of order deployments

env_borders = np.cumsum(env_counts)
detect[env_borders[:-1]] = 0  # disable false -1 on environment borders
unique_envs, _, env_counts = np.unique(envs_col, return_counts=True, return_index=True)

# distinguish the environments
detect *= np.repeat(np.arange(1, len(unique_envs) + 1), env_counts)
deltas, first_positions = np.unique(detect, return_index=True)
first_positions = first_positions[deltas < 0][::-1]

# find the nearest greater border index to each first occurrence position
last_positions = env_borders[np.searchsorted(env_borders, first_positions, side="right")]

# https://codereview.stackexchange.com/questions/83018/vectorized-numpy-version-of-arange-with-multiple-start-stop
invalid_counts = last_positions - first_positions
ranges = np.repeat(last_positions - invalid_counts.cumsum(), invalid_counts) + np.arange(invalid_counts.sum())

return notifications[DeploymentNotification.name.name].values[final_order[ranges]]
```

Note: there are no loops, everything is vectorized. It's easy for me to write this code because I am used to this... Please ask the details on Slack.

Reply:
black magic !
copied & pasted